v2/internal/crdmanagement/crd_installation_instruction.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package crdmanagement
+
+import (
+	"fmt"
+
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+type DiffResult string
+
+const (
+	NoDifference     = DiffResult("NoDifference")
+	SpecDifferent    = DiffResult("SpecDifferent")
+	VersionDifferent = DiffResult("VersionDifferent")
+)
+
+type FilterResult string
+
+const (
+	MatchedExistingCRD = FilterResult("MatchedExistingCRD")
+	MatchedPattern     = FilterResult("MatchedPattern")
+	Excluded           = FilterResult("Excluded")
+)
+
+type CRDInstallationInstruction struct {
+	CRD apiextensions.CustomResourceDefinition
+
+	// FilterResult contains the result of if the CRD was considered for installation or not
+	FilterResult FilterResult
+	// FilterReason contains a user-facing reason for why the CRD was or was not considered for installation
+	FilterReason string
+
+	// DiffResult contains the result of the diff between the existing CRD (if any) and the goal CRD. This may
+	// be nil if the CRD was filtered from consideration before the diff phase.
+	DiffResult DiffResult
+}
+
+func (i *CRDInstallationInstruction) ShouldApply() (bool, string) {
+	excluded := i.FilterResult == Excluded
+	if excluded {
+		return false, i.FilterReason
+	}
+
+	isSame := i.DiffResult == NoDifference
+	if isSame {
+		return false, i.DiffReason()
+	}
+
+	return true, i.DiffReason()
+}
+
+func (i *CRDInstallationInstruction) DiffReason() string {
+	switch i.DiffResult {
+	case NoDifference:
+		return fmt.Sprintf("No difference between existing and goal CRD %q", makeMatchString(i.CRD))
+	case SpecDifferent:
+		return fmt.Sprintf("The spec was different between existing and goal CRD %q", makeMatchString(i.CRD))
+	case VersionDifferent:
+		return fmt.Sprintf("The version was different between existing and goal CRD %q", makeMatchString(i.CRD))
+	default:
+		return fmt.Sprintf("Unknown DiffResult %q", i.DiffResult)
+	}
+}
```

Comment:
You could make this a method on `DiffResult` directly and avoid the type-envy smell:

```suggestion
func (d DiffResult) Reason(crd string) string {
	switch d {
	case NoDifference:
		return fmt.Sprintf("No difference between existing and goal CRD %q", crd)
	case SpecDifferent:
		return fmt.Sprintf("The spec was different between existing and goal CRD %q", crd)
	case VersionDifferent:
		return fmt.Sprintf("The version was different between existing and goal CRD %q", crd)
	default:
		return fmt.Sprintf("Unknown DiffResult %q", d)
	}
}
```